File: controller/e_meteor.cpp
//
// Created by Nico Wittig on 2019-03-03.
//

#include "e_meteor.hpp"

e_meteor::e_meteor(light_element *le, bool forward, uint8_t trail_decay) : effect(le, EFFECT_NAME) {
    this->forward = forward;
    this->trail_decay = trail_decay;
}

e_meteor::e_meteor(light_element *le, bool forward, uint8_t trail_decay, bool no_random) : effect(le, EFFECT_NAME) {
    this->forward = forward;
    this->trail_decay = trail_decay;
    this->no_random = no_random;
}


e_meteor::~e_meteor() {
}


bool e_meteor::init() {
    set_color(CHSV(190, 255, 0));
    return true;
}


bool e_meteor::run() {

    // catch exceptions
    meteor_position = modulo(meteor_position, le->get_num_leds());

    // fade out the trail
    for (int led = 0; led < le->get_num_leds(); led++) {
        if (no_random || random8()%2 == 0) {
            fade_var(&le->leds[led%le->get_num_leds()].val, 0, -trail_decay);
        }
    }

    // make the meteor full brightness
    le->leds[meteor_position].val = 255;

    // move meteor
    meteor_position += (forward) ? 1 : -1;

    return false;
}

File: server/e_meteor.py
import random

from color import Color
from effect import Effect


class E_Meteor(Effect):
    def __init__(self, light_element, color_hue=190, forward=True, trail_decay=60, no_random=False):
        super().__init__(light_element, "Meteor")

        self.__color_hue = color_hue
        self.__forward = forward
        self.__trail_decay = abs(trail_decay)
        self.__no_random = no_random

        self.__meteor_position = 0

    def init(self):
        super().set_color(Color((self.__color_hue, 255, 0)))
        return True

    def run(self):
        num_pixels = self._Effect__light_element.num_pixels
        self.__meteor_position %= num_pixels

        # fade out the trail
        for i in range(0, num_pixels):
            if self.__no_random or random.randrange(10) >= 5:
                super().get_pixel(i).val = super()._Effect__fade_var(super().get_pixel(i).val, 0, -self.__trail_decay)[1]

        # meteor with full brightness
        super().get_pixel(self.__meteor_position).val = 255
        # move meteor
        self.__meteor_position += 1 if self.__forward else -1

        return True
